### ISSUE TITLE
[i2c] Unexpected data in ACQ FIFO after deep sleep wakeup

### ISSUE THREAD

**Issue by cfrantz** [lowRISC/opentitan#18510]
### Description

I am observing unexpected data in the I2C ACQ FIFO after deep sleep wakeup on the FPGA.

Test scenario:
1. Get I2C configured in target mode.  Set the address filter to 0x33 and 0x70..0x73.
2. Configure the pinmux wakeup detector for any edge on IOA7 (I2C SDA).
3. Put the chip into deep sleep.
4. From the host, issue a I2C read transaction to address 0x22.
5. The chip wakes up from deep sleep; the test program re-initializes I2C.
6. Configure I2C for target mode and set the address filter as before (0x33, 0x70-0x73).
7. Observe that there is an entry in the ACQ FIFO: byte=0x45 for a Start condition (byte 0x45 == addr 0x22 + read).

When the chip is awake, if I try to issue a read (from the host) to I2C address 0x22, I _do not_ observe data captured into the ACQ FIFO.

Expectations:
1. I don't expect to capture anything into the ACQ FIFO when the chip is in deep sleep.
2. I also don't expect to capture an address that doesn't match the address filter.

Work-around:
If I reset the FIFOs after setting the address filter and enabling target mode, I2C behaves as expected.

Images:
Wakeup transaction to 0x22 followed by successful read from 0x33.
![image](https://github.com/lowRISC/opentitan/assets/628742/f13a0ccf-7829-4705-b7bf-f41a02865e1a)

Wakeup transaction to 0x22 (zoomed):
![image](https://github.com/lowRISC/opentitan/assets/628742/060513de-c8b0-4d62-8497-285b652f426c)

**Comment by cfrantz** [comment]
#18484 can show this problem.

To see the values in the fifo after deep sleep, change `configure_device_address` to not call `reset_fifos`, but rather use the DIF functions to read bytes out of the ACQ FIFO.

**Comment by msfschaffner** [comment]
@cfrantz I checked the I2C clock and reset signals behavior during a deep sleep power cycle using the `chip_sw_pwrmgr_deep_sleep_all_wake_ups` test, and they both seem to be shut off as expected.

Are we maybe interrupting the sleep entry process by sending the I2C transaction too early?
We can check that by reading the power manager `WAKE_INFO` register...

**Comment by msfschaffner** [comment]
Ok from our offline discussions it looks like this is not due to a fall-through scenario.

The other thing that comes to my mind after inspecting a deep-sleep power simulation is that the FPGA does not model any of the power-up delays correctly.

I.e., in deep sleep, the slow power manager FSM sits in `SlowPwrStateLowPower` waiting for a wakeup request. Once received, it cycles through a couple of states that basically just wait for "power OK" and "all clocks OK" indications before turning on the I2C clock (it doesn't wait for OTP initialization before turning on the I2C clock). These rampup delays are modelled in DV but not on FPGA. 

That means that the I2C clock may potentially come on *very* quickly when a transition on SDA is seen (within just a few AON clock cycles, each one being 1/250kHz = 4us on the CW310). 

Looking at the waveform above it looks like it takes something in ther order of 10us from the point where SDA goes low until the first SCL clock edge - so it still seems a bit tight. But this could explain the symptoms you are seeing.

We could verify this behavior using an integrated logic analyzer probe (ILA) - but I don't think this is critical for the M2.5.1 release.

**Comment by msfschaffner** [comment]
@moidx I moved this into M2.5.2. Please let me know if you disagree.

**Comment by moidx** [comment]
@cfrantz to document workaround in release notes for M2.5.1-RC0.

**Comment by johngt** [comment]
Ping - for release notes update. @cfrantz / @moidx

**Comment by msfschaffner** [comment]
The workaround has been noted in the release notes for RC0.
I am moving this to M2.5 Backlog for now.

@cfrantz @moidx how should we proceed with this after M2.5?
Should we add an I2C enable bit that makes sure the I2C only accepts new data if the configuration has been explicitly enabled?

**Comment by a-will** [comment]
> The workaround has been noted in the release notes for RC0. I am moving this to M2.5 Backlog for now.
> 
> @cfrantz @moidx how should we proceed with this after M2.5? Should we add an I2C enable bit that makes sure the I2C only accepts new data if the configuration has been explicitly enabled?

These already exist, don't they? https://github.com/lowRISC/opentitan/blob/23969736eba95bb42b472feb3d7432ed99f21a63/hw/ip/i2c/data/i2c.hjson#L121-L133

**Comment by msfschaffner** [comment]
Right, let me rephrase: we probably want to double check why the I2C ACQ FIFO accepts data right after the I2C module has been reset in that case, since `ENABLEHOST` and `ENABLETARGET` should both be 0 at that point.

It may very well be that everything is working as expected and the "unexpected" data makes it into the ACQ FIFO just because the system wakes up and re-initializes very quickly compared to the I2C waveform that is driven to wake up the system.

**Comment by msfschaffner** [comment]
@cfrantz I believe everything is working as intended here, based on the discussion above. OK if we close this issue, or would you like to do some more testing?
